### PR TITLE
fix: ButtonBase for flex layouts

### PR DESCRIPTION
**Pull request by georgewrmarshall**
## **Description**

Refactors ButtonBase and ButtonAnimated to properly support flex layouts in React Native.

### What is the reason for the change?

ButtonBase components couldn't use `flex-1` in flex row layouts due to nested wrapper architecture in ButtonAnimated. The `w-auto` width constraint also prevented proper flex behavior.

### What is the improvement/solution?

1. **Simplified ButtonAnimated**: Uses `Animated.createAnimatedComponent(Pressable)` (68 lines vs 111)
2. **Refactored ButtonBase**: Uses `tw.style()` for better readability and removed content wrapper View
3. **Fixed width behavior**: Changed from `w-auto` to `self-start` for proper flex support
4. **Updated documentation**: Clear guidance on horizontal vs vertical flex layouts

## **Related issues**

Fixes: (internal issue)

## **Manual testing steps**

1. Run Storybook: `yarn storybook:ios` or `yarn storybook:android`
2. Navigate to Components > ButtonBase
3. Test **FlexLayout** story - buttons should share space equally with `flex-1`
4. Test **isFullWidth** story - first button content-width, second full-width
5. Test **Sizes** story - verify all size variants render correctly
6. Test **IsLoading** story - verify loading state and spinner
7. Test button press animations - should scale to 0.97 on press
8. Test all Button variant stories (ButtonPrimary, ButtonSecondary, ButtonTertiary)

### **Before**

Buttons couldn't use `flex-1` in horizontal layouts - would either overflow or not stretch properly.

https://github.com/user-attachments/assets/bd19db75-7004-4401-a3ef-b33d6a4deeff

### **After**

Buttons properly support flex layouts with `twClassName="flex-1"` in horizontal rows and `isFullWidth` in vertical columns.

https://github.com/user-attachments/assets/6fc85f9b-0bdd-4282-ac7e-f91e1283fa0c

Also tested the [preview package](https://github.com/MetaMask/metamask-design-system/pull/870#issuecomment-3881746605)and checked in mobile for `ButtonHero` which uses `ButtonBase` under the hood. See [branch here](https://github.com/MetaMask/metamask-mobile/commit/a3cfc395c52b9e56070b6a278aa7c44ba631470d)

https://github.com/user-attachments/assets/ad73b956-7e6e-4909-bd0f-105813193e23

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs)
- [x] I've completed the PR template to the best of my ability
- [x] I've included tests if applicable
- [x] I've documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I've applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

---

> [!NOTE]
> **Medium Risk**
> Touches core button layout and press animation plumbing (`ButtonBase`/`ButtonAnimated`), which can subtly affect sizing, transforms, and loading visuals across apps. Changes are well-covered by updated stories/docs and expanded unit tests, but UI regressions are still possible.
> 
> **Overview**
> Fixes `ButtonBase` layout/animation issues that prevented reliable use in flex row layouts (e.g., side-by-side buttons with `twClassName="flex-1"`).
> 
> `ButtonAnimated` is reworked to be an `AnimatedPressable`, track pressed state internally, and *merge user-provided `transform` styles* with the press-scale animation instead of overwriting them. `ButtonBase` updates its container styling (notably using `self-start` when not `isFullWidth`) and changes loading rendering to show the spinner as an overlay while hiding individual content elements via `opacity-0` (including icons/accessories and both string and ReactNode children) without relying on a single content wrapper.
> 
> Storybook examples and README are updated with explicit guidance for horizontal vs vertical layouts, and tests are expanded/adjusted across `ButtonBase` and button variants to cover function-based classNames, loading-state opacity, custom accessories/children, and custom transforms.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 52fc508a57cebe1c431a86657e3073e69dffa44e. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: packages/design-system-react-native/src/components/ButtonBase/ButtonBase.stories.tsx [hunk 1]
```diff
@@ -130,6 +130,15 @@ export const WithLongText: Story = {
...
+    </Box>
+  ),
+};
+
+export const FlexLayout: Story = {
+  render: () => (
+    <Box flexDirection={BoxFlexDirection.Row} gap={4}>
+      <ButtonBase twClassName="flex-1">Lorem ipsum</ButtonBase>
+      <ButtonBase twClassName="flex-1">Lorem ipsum</ButtonBase>
+    </Box>
```
Comment (georgewrmarshall) [inline packages/design-system-react-native/src/components/ButtonBase/ButtonBase.stories.tsx]: Adding flex layout story to check flex layout

File: packages/design-system-react-native/src/components/ButtonBase/ButtonBase.tsx [hunk 2]
```diff
@@ -133,65 +140,73 @@ export const ButtonBase = ({
...
+            textProps={{
+              variant: TextVariant.BodyMd,
+              fontWeight: FontWeight.Medium,
+              color: TextColor.TextDefault,
+              twClassName: `shrink grow-0 flex-wrap text-center ${isLoading ? 'opacity-0' : ''} ${textClassName ? textClassName(pressed) : ''}`,
+              ...textProps,
+            }}
+          >
+            {children}
+          </TextOrChildren>
```
[the remark this answers is not in the document]
Comment (georgewrmarshall) [reply packages/design-system-react-native/src/components/ButtonBase/ButtonBase.tsx]: Fixed here https://github.com/MetaMask/metamask-design-system/pull/870/commits/5e2ed42b3dc6196f8c010375d0234732f3d1c81b

temporary story

```
<ButtonBase isLoading loadingText="With Loading Text">
  <Text>ButtonBase</Text>
</ButtonBase>
```

### Before

Screenshot 2026-02-11 at 5 18 28 PM

### After

Screenshot 2026-02-11 at 5 33 29 PM

File: packages/design-system-react-native/src/components/ButtonBase/ButtonBase.tsx [hunk 3]
```diff
@@ -131,65 +140,77 @@ export const ButtonBase = ({
...
-            }`}
-            testID="content-container"
-          >
-            {finalStartIconName ? (
-              <Icon
-                name={finalStartIconName}
-                size={IconSize.Sm}
-                twClassName={`shrink-0 ${iconClassName ? iconClassName(pressed) : ''}`}
-                {...startIconProps}
+          {/* Loading spinner overlay */}
```
Comment (georgewrmarshall) [inline packages/design-system-react-native/src/components/ButtonBase/ButtonBase.tsx]: Spinner now only renders when isLoading is true. Previously it was always in the DOM but hidden with opacity-0, which could cause unnecessary re-renders.

File: packages/design-system-react-native/src/components/ButtonBase/ButtonBase.tsx [hunk 4]
```diff
@@ -131,65 +140,77 @@ export const ButtonBase = ({
...
+            />
+          ) : (
+            startAccessory && (
+              <View style={tw.style(isLoading && 'opacity-0')}>
+                {startAccessory}
+              </View>
+            )
+          )}
+
+          {typeof children === 'string' ? (
```
Comment (georgewrmarshall) [inline packages/design-system-react-native/src/components/ButtonBase/ButtonBase.tsx]: Fixes existing bug where non-string React element children would remain visible during loading. TextOrChildren only applies textProps to strings, so non-string children now get wrapped in a View with opacity control.

File: packages/design-system-react-native/src/components/Button/variants/ButtonPrimary/ButtonPrimary.test.tsx [hunk 5]
```diff
@@ -211,16 +209,18 @@ describe('ButtonPrimary', () => {
...
     expect(spinnerStyles).toStrictEqual(
       expect.arrayContaining([expect.objectContaining(tw`${spinnerTW}`)]),
     );
 
-    const content = getByTestId('content-container');
-    const contentStyles = flattenStyles(content.props.style);
-    expect(contentStyles).toStrictEqual(
-      expect.arrayContaining([expect.objectContaining(tw`${contentTW}`)]),
+    // Verify content is hidden with opacity-0
+    const text = getByText('Loading');
```
Comment (georgewrmarshall) [inline packages/design-system-react-native/src/components/Button/variants/ButtonPrimary/ButtonPrimary.test.tsx]: Test updated to verify individual opacity control. Now checks that text element has opacity-0 during loading instead of checking the removed content-container wrapper.

File: packages/design-system-react-native/src/components/temp-components/ButtonAnimated/ButtonAnimated.tsx [hunk 6]
```diff
@@ -10,20 +10,27 @@ import Animated, {
...
+
 export const ButtonAnimated = ({
   onPressIn,
   onPressOut,
   disabled,
+  style,
+  children,
   ...props
 }: ButtonAnimatedProps) => {
+  const [isPressed, setIsPressed] = useState(false);
```
Comment (georgewrmarshall) [inline packages/design-system-react-native/src/components/temp-components/ButtonAnimated/ButtonAnimated.tsx]: Added pressed state tracking to enable function-based style and children props. This state is passed to style and children functions, allowing ButtonBase to conditionally apply styles based on press state.

File: packages/design-system-react-native/src/components/ButtonBase/ButtonBase.test.tsx [hunk 7]
```diff
@@ -319,27 +385,127 @@ describe('ButtonBase', () => {
...
+    const unpressedContent = childrenFn({ pressed: false });
+    const pressedContent = childrenFn({ pressed: true });
+
+    expect(unpressedContent).toBeDefined();
+    expect(pressedContent).toBeDefined();
+  });
+
+  it('applies textClassName when provided', () => {
+    const textClassNameFn = (pressed: boolean) =>
+      pressed ? 'text-pressed' : 'text-default';
```
Comment (georgewrmarshall) [inline packages/design-system-react-native/src/components/ButtonBase/ButtonBase.test.tsx]: Added comprehensive test coverage for function-based textClassName and iconClassName in both normal and loading states. These tests ensure the new pressed state functionality works correctly across all scenarios.

File: packages/design-system-react-native/src/components/temp-components/ButtonAnimated/ButtonAnimated.tsx [hunk 8]
```diff
@@ -10,20 +10,55 @@ import Animated, {
...
+
+  // Merge user transforms with animation scale
+  const animatedStyle = useAnimatedStyle(() => {
+    'worklet';
     return {
-      transform: [{ scale: animation.value }],
+      transform: [
+        ...userTransforms, // User transforms first (rotate, translate, etc.)
+        { scale: animation.value }, // Then animation scale
+      ],
```
[the remark this answers is not in the document]
Comment (georgewrmarshall) [reply packages/design-system-react-native/src/components/temp-components/ButtonAnimated/ButtonAnimated.tsx]: This is the desired behavior - we want the button animation to always work, even if consumers provide custom styles. If consumers need to completely override the animation, they should use a plain Pressable instead of ButtonAnimated

File: packages/design-system-react-native/src/components/ButtonBase/ButtonBase.tsx [hunk 9]
```diff
@@ -131,65 +140,77 @@ export const ButtonBase = ({
...
-                size={IconSize.Sm}
-                twClassName={`shrink-0 ${iconClassName ? iconClassName(pressed) : ''}`}
-                {...endIconProps}
-              />
-            ) : (
-              endAccessory
-            )}
-          </View>
+          ) : (
+            <View style={tw.style(isLoading && 'opacity-0')}>{children}</View>
```
Comment (georgewrmarshall) [inline packages/design-system-react-native/src/components/ButtonBase/ButtonBase.tsx]: Good catch! This PR actually **fixes** an existing bug where non-string React element children would remain visible during loading.

**What changed:**
- Before: All content wrapped in a single View with opacity control - but TextOrChildren component ignored this for React element children
- After: Opacity applied individually to each child element, ensuring ALL children (string and non-string) are properly hidden during loading

**The fix:**
- String children: TextOrChildren applies `opacity-0` via twClassName
- React element children: Wrapped in View with `tw.style(isLoading && 'opacity-0')`

**About the style override concern:**
While technically consumer-provided style.opacity could override the hiding, this is an edge case we're accepting. The proper fix would add significant complexity (extracting and merging opacity), which we decided isn't worth it for ButtonBase's intended use case.

Test coverage added in ButtonBase.test.tsx to verify non-string children are hidden during loading.
